doc/CodeSize.md

```diff
@@ -0,0 +1,260 @@
+# Minimizing Tock code size
+
+Many embedded applications are ultimately limited by the flash space available on
+the board in use. This document provides tips on how to write Rust code such that
+it does not require an undue amount of flash, and highlights some options which
+can be used to reduce the size required for a particular image.
+
+## Code Style: tips for keeping Rust code small
+
+#### When to use generic types with trait bounds versus trait objects (`dyn`)
+
+Polymorphic structs and functions are one of the biggest sources of bloat in Rust
+binaries -- use of generic types can lead to bloat from monomorphization, while
+use of trait objects introduces vtables into the binary and limits opportunities
+for inlining.
+
+Use `dyn` when the function in question will be called with multiple concrete types;
+otherwise code size is increased for every concrete type used (monomorphization).
+
+```rust
+fn set_gpio_client(&dyn GpioClientTrait) -> Self {//...}
+
+// elsewhere
+let radio: Radio = Radio::new();
+set_gpio_client(&radio);
+
+let button: Button = Button::new();
+set_gpio_client(&button);
+```
+
+Use generics with trait bounds when the function is only ever called with a single public type;
+This reduces code size and run time cost. This increases source code complexity for decreased
+image size and decreased clock cycles used.
```

Comment:
The wording "when the function is only ever called with a single public type" may be confusing. The "single public type" part of the phrase is per board (or per binary in the case of `libtock-rs` apps). E.g. if Imix calls it with type `A` and Hail calls it with type `B` it should still be generic argument rather than a `dyn` argument.

Reply:
I added "per board"